FEXCore/Source/Interface/Core/OpcodeDispatcher/AVX_128.cpp

```diff
@@ -2607,4 +2610,98 @@ void OpDispatchBuilder::AVX128_VPGATHER(OpcodeArgs) {
   AVX128_StoreResult_WithOpSize(Op, Op->Src[1], ZeroPair);
 }
 
+void OpDispatchBuilder::AVX128_VCVTPH2PS(OpcodeArgs) {
+  const auto DstSize = GetDstSize(Op);
+  const auto SrcSize = DstSize / 2;
+  const auto Is128BitSrc = SrcSize == Core::CPUState::XMM_SSE_REG_SIZE;
+  const auto Is128BitDst = DstSize == Core::CPUState::XMM_SSE_REG_SIZE;
+
+  RefPair Src {};
+  if (Op->Src[0].IsGPR()) {
+    Src = AVX128_LoadSource_WithOpSize(Op, Op->Src[0], Op->Flags, !Is128BitSrc);
+  } else {
+    // In the event that a memory operand is used as the source operand,
+    // the access width will always be half the size of the destination vector width
+    // (i.e. 128-bit vector -> 64-bit mem, 256-bit vector -> 128-bit mem)
+    const auto SrcLoadSize = SrcSize;
+    Src.Low = LoadSource_WithOpSize(FPRClass, Op, Op->Src[0], SrcLoadSize, Op->Flags);
+  }
+
+  auto TransformLow = [this](Ref Src) -> Ref {
+    return _Vector_FToF(OpSize::i128Bit, OpSize::i32Bit, Src, OpSize::i16Bit);
+  };
+
+  auto TransformHigh = [this](Ref Src) -> Ref {
+    return _VFCVTL2(OpSize::i128Bit, OpSize::i16Bit, Src);
+  };
+
+  RefPair Result {};
+  Result.Low = TransformLow(Src.Low);
+
+  if (Is128BitSrc) {
+    Result.High = TransformHigh(Src.Low);
+  }
+
+  if (Is128BitDst) {
+    Result = AVX128_Zext(Result.Low);
+  }
+
+  AVX128_StoreResult_WithOpSize(Op, Op->Dest, Result);
+}
+
+void OpDispatchBuilder::AVX128_VCVTPS2PH(OpcodeArgs) {
+  const auto SrcSize = GetSrcSize(Op);
+  const auto Is128BitSrc = SrcSize == Core::CPUState::XMM_SSE_REG_SIZE;
+  const auto StoreSize = Op->Dest.IsGPR() ? OpSize::i128Bit : SrcSize / 2;
+
+  const auto Imm8 = Op->Src[1].Literal();
+  const auto UseMXCSR = (Imm8 & 0b100) != 0;
+
+  auto Src = AVX128_LoadSource_WithOpSize(Op, Op->Src[0], Op->Flags, !Is128BitSrc);
+
+  RefPair Result {};
+  auto TransformLow = [this](Ref Src) -> Ref {
+    return _Vector_FToF(OpSize::i128Bit, OpSize::i16Bit, Src, OpSize::i32Bit);
+  };
+
+  auto TransformHigh = [this](Ref Lower, Ref Src) -> Ref {
+    return _VFCVTN2(OpSize::i128Bit, OpSize::i32Bit, Lower, Src);
+  };
+
+  if (UseMXCSR) {
```

Comment:
Might be profitable to split up as

```
Ref OldFPCR;
if (!UseMXCSR) {
   Push()
}

Transform()

if (!Use) {
   Pop()
}
```

Then the transform lambdas are used only once not twice. And then we inline the lambdas and get simpler code overall, I think.

Reply:
Nice idea